modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java

```diff
@@ -120,41 +125,54 @@ public OperationMessageProcessor(ExtensionModel extensionModel,
     this.target = target;
     this.entityMetadataMediator = new EntityMetadataMediator(operationModel);
     this.policyManager = policyManager;
+    operationIdentifier = new ComponentIdentifier.Builder()
+        .withName(operationModel.getName())
+        .withNamespace(extensionModel.getName().toLowerCase())
+        .build();
+    classLoader = getExtensionClassLoader();
   }
 
   @Override
   public Event process(Event event) throws MuleException {
-    return withContextClassLoader(getExtensionClassLoader(), () -> {
-      Optional<ConfigurationInstance> configuration = getConfiguration(event);
+    try {
+      return just(event).transform(this).subscribe().block();
+    } catch (Exception e) {
+      throw rxExceptionToMuleException(e);
+    }
+  }
 
-      ComponentIdentifier operationIdentifier =
-          new ComponentIdentifier.Builder().withName(operationModel.getName())
-              .withNamespace(extensionModel.getName().toLowerCase()).build();
+  @Override
+  public Publisher<Event> apply(Publisher<Event> publisher) {
+    return from(publisher).concatMap(checkedFunction(event -> withContextClassLoader(classLoader, () -> {
+      Optional<ConfigurationInstance> configuration = getConfiguration(event);
 
       Map<String, Object> operationParameters = resolverSet.resolve(event).asMap();
-
-      OperationExecutionFunction operationExecutionFunction = (parameters, operationEvent) -> {
-        ExecutionContextAdapter operationContext = createExecutionContext(configuration, parameters, operationEvent);
-        MuleEvent muleEvent = doProcess(event, operationContext);
-        return (Event) muleEvent;
-      };
-
-      OperationPolicy policy =
-          policyManager.createOperationPolicy(operationIdentifier, event, operationParameters, operationExecutionFunction);
-      return policy.process(event);
+      ExecutionContextAdapter operationContext = createExecutionContext(configuration, operationParameters, event);
+
+      if (operationModel.isBlocking()) {
+        OperationExecutionFunction operationExecutionFunction =
+            (parameters, operationEvent) -> doProcess(operationEvent, operationContext).block();
+        OperationPolicy policy =
+            policyManager.createOperationPolicy(operationIdentifier, event,
+                                                operationParameters,
+                                                operationExecutionFunction);
+        return justOrEmpty(policy.process(event));
```

Comment:
Is null an expected result here?  Doing this will drop the event, and no callback will be invoked handing the source.  You can avoid this by using `nullSafeMap` which completes the eventContext.